### DISCUSSION TITLE
S2 Mini support

### DISCUSSION THREAD

**Discussion by phoddie** [phoddie/node-red-mcu#114] [General]
This is a [continuation of a discussion](https://github.com/phoddie/node-red-mcu/discussions/111#discussioncomment-5666591) started by @colinl about support for the S2 Mini development board.

The Moddable SDK has supported the ESP32-S2 MCU on the S2 Mini board [for some time](https://blog.moddable.com/blog/espidf42/). All the boards supported have used the usual serial to USB bridge chip to communicate with the development computer for flashing the firmware and debugging. The S2 Mini has only a USB port and so requires some additional support.

Over the weekend, I added USB support to our ESP32-S2 port, migrating over the support already in place for ESP32-S3. @mkellner advised on that and added some improvements to the build to make that a bit smoother. The USB support works well. Honestly, it is more convenient to work with devices that have serial. USB-only devices require manual button pushes to enter and exit programming mode, which is inconvenient. Alas, that's just the way it is.

Everything works nicely, except.... Wi-Fi! Wi-Fi scanning works as expected. But, connecting to Wi-Fi is not reliable from the S2 Mini. On other ESP32-S2 boards Wi-Fi works just fine, so we don't believe this is a software issue on our side. We also tested with an Adafruit QtPyS2, which is quite similar to the S2 Mini, and had no problem with Wi-Fi. As it turns out, there are more than a few reports of Wi-Fi related issues with the S2 Mini.  The workarounds include putting a coin on one edge of the board (!!) and soldering on a capacitor. It may be that the boards I received are from a bad batch.

So... our current plan is to push the changes for USB for S2 Mini in the next day or so. Maybe it will even work. ;) But between the manual intervention when programming via USB and the potential Wi-Fi issues, please do be prepared for some exploration.

**Comment by bartbutenaers** [comment]
Hi @phoddie,

Not an expert on this as you know...
But I had also installed my first wemo D1 mini last year, running EspEasy on top of it.  I had installed it in my basement, about 6 meters from my access point.  The wifi connection of my mini was incredible bad: the connection went gone all the time, and when I had a connection the web ui loaded very slow.

Somebody from the Node-RED community (dynamicdave) shared with me this information:

![image](https://user-images.githubusercontent.com/14224149/234185660-0c86e9bb-c3af-45a0-ba51-ab4e2c4bcb77.png)

After applying these settings in EspEasy, I had no problems anymore with my wifi.  

Don't know if this is of any help to you, but perhaps it gives you some ideas...

Good luck with this!
Bart

**Reply by phoddie** [reply]
Thanks, @bartbutenaers! I suspected that that low transmit (TX) power might be an issue. I see failures even with the S2 Mini a few feet away from the access point, which seems crazy. I'll try some of this out.

**Reply by phoddie** [reply]
I read the code in EspEasy to understand how the UI maps to ESP-IDF calls. It looks like there are three relevant APIs for the above. I added those locally at the end of `initWiFi`, [after the call to `esp_wifi_start`](https://github.com/Moddable-OpenSource/moddable/blob/73f318596575cffaf8d02d869c9f64f54033cefc/modules/network/wifi/esp32/xs6wifi.c#L533). (The `modLog` calls are more-or-less lightweight printf for portability - here there are used to confirm that no error is reported by the APIs)

```
	int8_t power = 0;
	int err = esp_wifi_get_max_tx_power(&power);
	modLog("esp_wifi_get_max_tx_power err power");
	modLogInt(err);
	modLogInt(power);

	err = esp_wifi_set_max_tx_power(84);
	modLog("esp_wifi_set_max_tx_power err");
	modLogInt(err);

	err = esp_wifi_set_protocol(ESP_IF_WIFI_STA, WIFI_PROTOCOL_11B | WIFI_PROTOCOL_11G);
	modLog("esp_wifi_set_protocol err");
	modLogInt(err);

	err = esp_wifi_set_ps(WIFI_PS_NONE);
	modLog("esp_wifi_set_ps err");
	modLogInt(err);
```

The default power is 78 so this bumps it marginally. Alas, there's no noticeable difference in behavior. I suspect that the boards I may not be transmitting. We also tried soldering on a capacitor as recommend in some posts, but no difference. It may well be that other boards work actually work. ;) We have had no Wi-Fi problems with other ESP32-S2 based boards, even those using USB.

**Reply by colinl** [reply]
If you commit what you have I will give it a go with mine.

**Reply by phoddie** [reply]
Thanks! I'm putting together some notes right now so you can do that. That should be up shortly on this thread.

**Comment by phoddie** [comment]
The latest ESP32-S2 support is committed. That includes a subplatform build target for the S2 Mini board (`esp32/s2mini`). To try it, grab the latest Moddable SDK. Here's what I do to flash the S2 Mini:

- Attached the S2 Mini via USB-C
- Put the board into programming mode. This is done by holding down the "0" button and quickly tapping the "RST" button. Note that my luck in doing this is far short of 100%.
- Use mcconfig to build as usual:
```
cd $MODDABLE/examples/network/http/httpget
mcconfig -d -m -p esp32/s2mini ssid=Moddable password=$MY_PASSWORD
```
- If the device **is** in programming mode, you'll see something like this when it starts to flash:
```
esptool.py esp32s2 -p /dev/cu.usbmodem01 -b 921600 --before=default_reset --after=no_reset write_flash --flash_mode dio --flash_freq 80m --flash_size 4MB 0x1000 bootloader/bootloader.bin 0x10000 xs_esp32.bin 0x8000 partition_table/partition-table.bin
esptool.py v3.3.2
Serial port /dev/cu.usbmodem01
Connecting...
Chip is ESP32-S2FNR2
Features: WiFi, Embedded Flash 4MB, Embedded PSRAM 2MB, ADC and temperature sensor calibration in BLK2 of efuse V2
```
- If the S2 Mini **is not** in programming mode, you'll see something like this instead
```
Successfully created esp32s2 image.
Generated /Users/hoddie/Projects/moddable/build/tmp/esp32/s2mini/debug/httpget/xsProj-esp32s2/build/xs_esp32.bin
[9/9] cd /Users/hoddie/Projects/moddable/build/tmp/esp32/s2mini/debug/httpge...dable/build/tmp/esp32/s2mini/debug/httpget/xsProj-esp32s2/build/xs_esp32.bi
xs_esp32.bin binary size 0xcb000 bytes. Smallest app partition is 0x3f0000 bytes. 0x325000 bytes (80%) free.

Project build complete. To flash, run this command:
/Users/hoddie/.espressif/python_env/idf4.4_py3.10_env/bin/python ../../../../../../../../../esp32/esp-idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after no_reset --chip esp32s2  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/xs_esp32.bin
or run 'idf.py -p (PORT) flash'
** Please set the UPLOAD_PORT environment variable **
Executing action: flash
Serial port /dev/cu.Bluetooth-Incoming-Port
Connecting.......................No serial ports found. Connect a device, or use '-p PORT' option to set a specific port.

/dev/cu.Bluetooth-Incoming-Port failed to connect: Failed to connect to Espressif device: No serial data received.
For troubleshooting steps visit: https://docs.espressif.com/projects/esptool/en/latest/troubleshooting.html
make: *** [all] Error 2
```
- After flashing is complete, the build will pause waiting for the device to reboot in run mode, so it can connect to xsbug. Here's what you see in the console when it is waiting:
```
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (83 compressed) at 0x00008000 in 0.1 seconds (effective 341.1 kbit/s)...
Hash of data verified.

Leaving...
Staying in bootloader.
 ESP32-S2
Running ninja in directory /Users/hoddie/Projects/moddable/build/tmp/esp32/s2mini/debug/httpget/xsProj-esp32s2/build
Executing "ninja flash"...
Done
```
- Press the "RST" button without holding down "D0". You'll see this in the console and then xsbug should connect.
```
product/vendor match: /dev/cu.usbmodem1234561
```

Good luck!

**Reply by phoddie** [reply]
Assuming the S2 Mini eventually behaves for someone.... the Wi-Fi headache distracted me so much that I overlooked that the S2 Mini has 2 MB of PSRAM. I got that working this morning and changed the default virtual machine partition to use some of that. That will make the runtime less likely to hit out-of-memory failures. That change should be live in the next few days.

**Reply by colinl** [reply]
I am visiting family so don't have many opportunities to play. I gave it a quick go last night and could not get it to recognise that it was in programming mode. Though if I tailed syslog while I pressed the button I did see it mounting as  /dev/ttyACM0. Which I assumed meant that it was ready. Hopefully will have another go later.

**Reply by colinl** [reply]
I am visiting family so don't have many opportunities to play. I gave it a quick go last night and could not get it to recognise that it was in programming mode. Though if I tailed syslog while I pressed the button I did see it mounting as  /dev/ttyACM0. Which I assumed meant that it was ready. Hopefully will have another go later.

**Reply by mkellner** [reply]
Hello @colinl , I understand you are busy, but when you return to this, try the following:

In the file `$MODDABLE/build/devices/esp32/targets/s2mini/manifest.json`, add two lines to the `build` section with `PROGRAMMING_VID` and `PROGRAMMING_PID` like:

```
	"build":{
		"ESP32_SUBCLASS": "esp32s2",
		"USE_USB": "1",
		"PROGRAMMING_VID": "303a",
		"PROGRAMMING_PID": "0002",
		"SDKCONFIGPATH": "./sdkconfig"
	},
```

The defaults in `make.esp32.mk` are for the ESP32-S3. That is, `PROGRAMMING_PID` for the ESP-S3 is `1001`.

The Linux build uses the VID and PID to find the USB port that the device is connected to. The tool it uses is in `$MODDABLE/build/devices/esp32/config/findUSBLinux`. With the wrong PID it was failing to connect. Peter is using the MacOS build, so he didn't witness the problem.

**Reply by colinl** [reply]
Thanks for that.  The manifest change takes us forward, I think, now I am getting
```
esptool.py v3.3.2
Creating esp32s2 image...
Merged 2 ELF sections
Successfully created esp32s2 image.
Generated /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/xs_esp32.bin
[7/7] cd /home/colinl/Projects/moddabl...we03/xsProj-esp32s2/build/xs_esp32.bin
xs_esp32.bin binary size 0x16dc00 bytes. Smallest app partition is 0x3e0000 bytes. 0x272400 bytes (63%) free.

Project build complete. To flash, run this command:
/home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python ../../../../../../../../../esp32/esp-idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after no_reset --chip esp32s2  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/xs_esp32.bin
or run 'idf.py -p (PORT) flash'
Triggering serial port /dev/ttyACM0.
Executing action: flash
Serial port /dev/ttyACM0
Connecting...
Detecting chip type... Unsupported detection protocol, switching and trying again...
Detecting chip type...[1/5] cd /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/esp-idf/esptool_py && /home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python /home/colinl/esp32/esp-idf/components/partition_table/check_sizes.py --offset 0x8000 partition --type app /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/partition_table/partition-table.bin /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/xs_esp32.bin
xs_esp32.bin binary size 0x16dc00 bytes. Smallest app partition is 0x3e0000 bytes. 0x272400 bytes (63%) free.
[2/5] Performing build step for 'bootloader'
[1/1] cd /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/bootloader/esp-idf/esptool_py && /home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python /home/colinl/esp32/esp-idf/components/partition_table/check_sizes.py --offset 0x8000 bootloader 0x1000 /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/bootloader/bootloader.bin
Bootloader binary size 0x32e0 bytes. 0x3d20 bytes (55%) free.
[2/3] cd /home/colinl/esp32/esp-idf/components/esptool_py && /usr/bin/cmake -D IDF_PATH="/home/colinl/esp32/esp-idf" -D SERIAL_TOOL="/home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python /home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py --chip esp32s2" -D SERIAL_TOOL_ARGS="--before=default_reset --after=no_reset write_flash @flash_args" -D WORKING_DIRECTORY="/home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build" -P /home/colinl/esp32/esp-idf/components/esptool_py/run_serial_tool.cmake
esptool.py esp32s2 -p /dev/ttyACM0 -b 921600 --before=default_reset --after=no_reset write_flash --flash_mode dio --flash_freq 80m --flash_size 4MB 0x1000 bootloader/bootloader.bin 0x10000 xs_esp32.bin 0x8000 partition_table/partition-table.bin
esptool.py v3.3.2
Serial port /dev/ttyACM0
Connecting...
Traceback (most recent call last):
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 5482, in <module>
    _main()
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 5475, in _main
    main()
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 4782, in main
    esp = esp or get_default_connected_device(ser_list, port=args.port, connect_attempts=args.connect_attempts,
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 116, in get_default_connected_device
    _esp.connect(before, connect_attempts)
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 689, in connect
    last_error = self._connect_attempt(mode=mode, usb_jtag_serial=usb_jtag_serial, extra_delay=extra_delay)
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 639, in _connect_attempt
    self.bootloader_reset(usb_jtag_serial, extra_delay)
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 623, in bootloader_reset
    self._setDTR(False)  # IO0=HIGH, done
  File "/home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py", line 557, in _setDTR
    self._port.setDTR(state)
  File "/home/colinl/.espressif/python_env/idf4.4_py3.10_env/lib/python3.10/site-packages/serial/serialutil.py", line 603, in setDTR
    self.dtr = value
  File "/home/colinl/.espressif/python_env/idf4.4_py3.10_env/lib/python3.10/site-packages/serial/serialutil.py", line 473, in dtr
    self._update_dtr_state()
  File "/home/colinl/.espressif/python_env/idf4.4_py3.10_env/lib/python3.10/site-packages/serial/serialposix.py", line 715, in _update_dtr_state
    fcntl.ioctl(self.fd, TIOCMBIC, TIOCM_DTR_str)
OSError: [Errno 71] Protocol error
CMake Error at run_serial_tool.cmake:56 (message):
  /home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python
  /home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py --chip
  esp32s2 failed


FAILED: CMakeFiles/flash /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build/CMakeFiles/flash 
cd /home/colinl/esp32/esp-idf/components/esptool_py && /usr/bin/cmake -D IDF_PATH="/home/colinl/esp32/esp-idf" -D SERIAL_TOOL="/home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python /home/colinl/esp32/esp-idf/components/esptool_py/esptool/esptool.py --chip esp32s2" -D SERIAL_TOOL_ARGS="--before=default_reset --after=no_reset write_flash @flash_args" -D WORKING_DIRECTORY="/home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build" -P /home/colinl/esp32/esp-idf/components/esptool_py/run_serial_tool.cmake
ninja: build stopped: subcommand failed.
ninja failed with exit code 1
 ESP32-S2
Running ninja in directory /home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/xsProj-esp32s2/build
Executing "ninja flash"...
make: *** [/home/colinl/Projects/moddable/build/tmp/esp32/s2mini/debug/.node-red-we03/makefile:1037: all] Error 2
```

**Reply by mkellner** [reply]
It looks like the device is in a strange state. Can you unplug the device and try again?

**Reply by colinl** [reply]
I did try a couple of times, and once without putting into program mode in order to check it did something different. I will try again.

**Reply by colinl** [reply]
I have tried again with various combinations of unplugging and plugging and pressing the buttons and keep getting the same error.

**Reply by mkellner** [reply]
Hmmm. Could you please try building `helloworld` as described in the [Linux installation instructions](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#installing-2) to see if we can succeed on the `s2mini` with a simple case. It's basically:

```
xsbug
cd $MODDABLE/examples/helloworld
mcconfig -d -m -p esp32/s2mini
```

**Reply by mkellner** [reply]
I've installed node-red and see the problem.

The s2mini (or other esp32 devices using USB) presents itself as /dev/ttyACM0 for installation and /dev/ttyACM1 when it is running. The Moddable SDK's serial2xsbug can connect to a vid:pid, but I think there may be work necessary for that to work with the node-red-mcu plugin.

**Reply by colinl** [reply]
I am not using node-red-mcu-plugin, but I am using node-red-mcu.
Can I force the port in mcconfig?

**Reply by colinl** [reply]
> cd $MODDABLE/examples/helloworld
> mcconfig -d -m -p esp32/s2mini

I am seeing the same error when I run that.

**Comment by colinl** [comment]
However, after building, if I run 
```
cd $MODDABLE/build/tmp/esp32/s2mini/debug/helloworld/xsProj-esp32s2 &&
/home/colinl/.espressif/python_env/idf4.4_py3.10_env/bin/python ../../../../../../../../../esp32/esp-idf/components/esptool_py/esptool/esptool.py -p /dev/ttyACM0 -b 460800 --before default_reset --after no_reset --chip esp32s2  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/xs_esp32.bin
```
Then it does appear to download to the device :)

**Reply by colinl** [reply]
I can also manually download the build from my node-red flow, but I can't work out how to get xsbug to connect to it.

**Comment by phoddie** [comment]
FWIW – in case I was missing something obvious, I tried running one of the Wi-Fi examples from the ESP-IDF. That completely eliminates any code written by me. ;) Alas, it does not seem want to connect. I think I should get different S2 mini hardware.

**Reply by colinl** [reply]
This may be relevant - apparently a bad batch of devices. https://discourse.nodered.org/t/iipo-checker-for-up-to-6s/78427/14

**Reply by phoddie** [reply]
Thanks for sharing that. I'm pretty certain I have units from that bad batch. @mkellner purchased some S2 mini boards independently. Those work just fine with Wi-Fi at the Moddable office while mine fail there. So, actually, I think S2 mini support is at parity with other ESP32 dev boards at this point.

**Reply by colinl** [reply]
None of which explains why I can't download to the device, other than using the command line to manually download the file. Which is tedious.

**Reply by phoddie** [reply]
Sure. That is a tooling topic, not a porting topic. The command line tools are what the Moddable SDK provides and that works. If I understand you correctly, you are having problems downloading with the MCU plug-in. If that understanding is correct, we should open a discussion with @ralphwetzel. I'm happy to support that integration work, of course, I just don't think I can diagnose / solve it from purely from the Moddable SDK's perspective.

N.B. These "USB native" devices are more complicated for the tools to work than those which use a serial-to-USB bridge chip. They are increasingly common since they are smaller and less expensive to produce. We've done a bunch of work recently to make that easier and more reliable and expect to continue to refine that as needed.

**Reply by colinl** [reply]
I am not using node-red-mcu-plugin for this test, just node-red-mcu.

**Reply by phoddie** [reply]
My misunderstanding then. Apologies. Would you mind providing opening an issue in the Moddable SDK repository with the steps to reproduce the failure and what you are doing to flash? That would help us track and resolve this.

**Reply by colinl** [reply]
Will do. I will upgrade everything to the latest first though, and recheck, just in case.

**Reply by phoddie** [reply]
Thank you. Very much appreciated.
